src/app/shared/context.service.ts

```diff
@@ -14,59 +18,183 @@ import { Broadcaster } from '../shared/broadcaster.service';
-    // Find the most specific context menu path and display it
-    let c;
-    for (let m of this.dummy.contexts) {
-      if (this.router.url.startsWith(m.path)) {
-        if (c == null || m.path.length > c.path.length) {
-          c = m;
-        }
+  lookupContextType(type: string): ContextType {
+    let ct = this.dummy.CONTEXT_TYPES.get(type);
+    // Make a copy of the context type before returning
+    return ct ? JSON.parse(JSON.stringify(ct)) : null;
+  }
+
+  get default(): Context {
+    return this._default;
+  }
+
+  set current(context: Context) {
+    if (!context) {
+      context = this.default;
+    }
+    if (this._current !== context) {
+      // If the context changed
+      this._current = context;
+      this.addRecentContext(context);
+    }
+  }
+
+  private addRecentContext(context: Context) {
+    for (let i = this.recent.length - 1; i >= 0; i--) {
+      if (this.recent[i].path === context.path) {
+        this.recent.splice(i, 1);
       }
     }
-    this._current = c || this.dummy.defaultContext;
-    if (this.current.type.menus) {
-      for (let n of this.current.type.menus) {
-        // Build the fullPath if not already done
-        n.fullPath = n.fullPath || this.buildPath(this.current.path, n.path);
-        // Clear the menu's active state
-        n.active = false;
+    this.recent.unshift(context);
+    // Truncate the context
+    if (this.recent.length > this.RECENT_CONTEXT_LENGTH) {
+      this.recent.splice(
+        this.RECENT_CONTEXT_LENGTH,
+        this.recent.length - this.RECENT_CONTEXT_LENGTH
+      );
+    }
+    this.broadcaster.broadcast('save');
+  }
+
+  private buildContextMenus(context: Context) {
+    if (typeof context.type === 'string') {
+      context.type = this.lookupContextType(context.type);
+    }
+    if (context.type.menus) {
+      for (let n of context.type.menus) {
+        n.fullPath = this.buildPath(context.path, n.path);
         if (n.menus) {
           for (let o of n.menus) {
-            // Build the fullPath if not already done
-            o.fullPath = o.fullPath || this.buildPath(this.current.path, n.path, o.path);
-            // Clear the menu's active state
-            o.active = false;
-            if (o.fullPath === this.router.url) {
-              o.active = true;
-              n.active = true;
-            }
+            o.fullPath = this.buildPath(context.path, n.path, o.path);
           }
-        } else if (n.fullPath === this.router.url) {
-          n.active = true;
         }
       }
     }
   }
 
+  private computeContext(): Context {
+    // First, check if there is a recent context
+    let entityStr: string = this.extractEntity(this.router.url);
+    let spaceStr: string = this.extractSpace(this.router.url);
+    // TODO Implement team URLs
+    let teamStr: string = null;
+    // TODO This is horrible, make a switch?
+    for (let c of this.recent) {
+      if (entityStr && c.entity && !c.team && typeof c.entity === 'User') {
```

Comment:
I don't think this is the way "typeof" works. I think typeof is comparing with one of the standard types: "undefined", "object", "boolean", "number", "string", "function", "symbol".